### ISSUE TITLE
Perform automatic type conversion in expressions when unambiguous

### ISSUE THREAD

**Issue by stevage** [mapbox/mapbox-gl-js#9501]
## Motivation

Currently, expressions don't do any automatic type conversion. Except for colors, for some reason:

>Expressions perform only one kind of implicit type conversion: a data expression used in a context where a color is expected will convert a string representation of a color to a color value. In all other cases, if you want to convert between types, you must use one of the type conversion expression operators: to-boolean, to-number, to-string, or to-color. 

This means that expressions frequently become very verbose. For instance, I always have to wrap `['feature-state', 'selected']` in `to-boolean` so that undefined (never set) states get coerced to null.

Another example is given [here](https://stackoverflow.com/questions/61020660/how-to-write-a-mapbox-paint-expression-that-accounts-for-zoom-feature-state-an/61022285?noredirect=1#comment108033256_61022285) where my answer is almost readable, and the one with all the type conversions is...not.

For me, even though I write Mapbox-GL expressions all the time, I'm actually not super confident in knowing when I need one and when I don't. Often these errors don't show up until run-time. It's just a constant source of friction. (Another part of this problem is that my first attempt to perform the type conversion is often to use the `['boolean']`operator, which is wrong.)

Another case would be to automatically convert to number the arguments of `['<']` etc.

## Design

My suggestion is to perform automatic type conversion when it's unambiguous. For instance, the operators `!`, `all`, `any` and `case` all take a boolean (and only a boolean) argument. There doesn't seem to be anything stopping Mapbox-GL automatically inserting a `['to-boolean', ...]` around that argument.

I'm not sure what the downsides are. Potentially one could complain that now there is inconsistency, in some places you don't need type conversion, and some places you do. To my mind that's better than always needing it. (And obviously you can still write it if you prefer.) And that didn't seem to be a problem in the case of colors.

Or perhaps there is a small performance hit?

**Comment by FrenchYann** [comment]
I'm not very fan of automatic type conversion. Expressions are hard enough to debug, I'd rather not have potential errors hidden form me.
Though I agree with you about having to have to write type conversion code every time is a problem.
My solution is usually to either keep the part of the expression that are repeated in a variable, or use simple helper functions to make the expression itself more readable

In the case of the stackoverflow example, you could do something like this:

```javascript
const ex_not = (a)   => ['!',   ['to-boolean', a]];
const ex_or = (a, b) => ['any', ['to-boolean', a],
                                ['to-boolean', b]];
const ex_and = (a,b) => ['all', ['to-boolean', a],
                                ['to-boolean', b]];
const ex_not_hidden = ex_not(['feature-state', 'hidden']);

map.addLayer({
  id: 'places',
  type: 'circle',
  source: 'places',
  paint: {
    'circle-color': ['get', 'color'],
    'circle-opacity': [
      'step', ['zoom'],
          ['case', ex_and(ex_not_hidden, ['get', 'red']),
             0.25,
             0],
          10, [ 'case', ex_and(ex_not_hidden, ex_or(['get', 'red'], ['get', 'green'])),
               0.5,
               0 ],
          12, [ 'case', ex_not_hidden,
               0.75,
               0 ],
    ],
  }
});
```
I use the `ex_` prefix for all variables and functions that should evaluate to a valid mapbox expression

**Comment by stevage** [comment]
Yeah, I thought about something like that, but I think in this case, the cure is worse than the disease. Now we have a mix of mapbox-gl expressions, and actual JavaScript with arbitrary function names. Portability between different projects is going to be minimal, examples on StackOverflow won't work etc.

> I'm not very fan of automatic type conversion. Expressions are hard enough to debug, I'd rather not have potential errors hidden form me.

Given that type conversion is a big source of errors for me, I'm going to disagree on this one.
